### ISSUE TITLE
possible data race in ~grpc_chttp2_transport

### ISSUE THREAD

**Issue by madeleineth** [grpc/grpc#18211]
### What version of gRPC and what language are you using?

C++, gRPC 1.19.0
 
### What operating system (Linux, Windows,...) and version?

Ubuntu 18.10
 
### What runtime / compiler are you using (e.g. python version or version of gcc)

Clang 7.0.0-3
 
### What did you do?

1. Checked out gRPC v1.19.0 and built it with the Clang 7 that comes with Ubuntu 18.10 on amd64.
2. Built the source over here with `make`: https://github.com/madeleineth/grpc-race-repro (It is a trivial gRPC server and client built with ThreadSanitizer.)
3. Run `./server` in one terminal, and `./client` in another a few times.
 
### What did you expect to see?

In the terminal I ran `server`, I expected the string "Ping!" each time I ran the client.
 
### What did you see instead?

[This TSan dump](https://github.com/madeleineth/grpc-race-repro/blob/master/tsan-output.txt):
```
Ping!
==================
WARNING: ThreadSanitizer: data race (pid=10695)
  Write of size 8 at 0x7b540000ff08 by thread T5:
    #0 free ??:? (server+0x4489dd)
    #1 ~grpc_chttp2_transport /home/madeleineth/grpc/src/core/ext/transport/chttp2/transport/chttp2_transport.cc:171 (libgrpc.so.7+0xb7fce)

  Previous read of size 1 at 0x7b540000ff0f by thread T4:
    #0 sendmsg ??:? (server+0x433f4e)
    #1 tcp_send(int, msghdr const*) /home/madeleineth/grpc/src/core/lib/iomgr/tcp_posix.cc:580 (libgrpc.so.7+0x96070)
    #2 tcp_flush((anonymous namespace)::grpc_tcp*, grpc_error**) /home/madeleineth/grpc/src/core/lib/iomgr/tcp_posix.cc:895 (libgrpc.so.7+0x96070)
    #3 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallOpSendMessage, grpc::internal::CallOpServerSendStatus, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >::FillOps(grpc::internal::Call*) ??:? (server+0x4e3981)
    #4 grpc::internal::Call::PerformOps(grpc::internal::CallOpSetInterface*) ??:? (server+0x4cd507)
    #5 grpc::internal::RpcMethodHandler<repro::ReproService::Service, repro::EmptyMessage, repro::EmptyMessage>::RunHandler(grpc::internal::MethodHandler::HandlerParameter const&) ??:? (server+0x4e26dc)
    #6 grpc::Server::SyncRequest::CallData::ContinueRunAfterInterception() /home/madeleineth/grpc/src/cpp/server/server_cc.cc:306 (libgrpc++_unsecure.so.1+0x56e1d)

  Thread T5 'grpcpp_sync_ser' (tid=10718, running) created by thread T4 at:
    #0 pthread_create ??:? (server+0x42e0c6)
    #1 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*) /home/madeleineth/grpc/src/core/lib/gprpp/thd_posix.cc:73 (libgrpc.so.7+0x12bfbc)
    #2 grpc_core::(anonymous namespace)::ThreadInternalsPosix* grpc_core::New<grpc_core::(anonymous namespace)::ThreadInternalsPosix, char const*&, void (*&)(void*), void*&, bool*>(char const*&, void (*&)(void*), void*&, bool*&&) /home/madeleineth/grpc/./src/core/lib/gprpp/memory.h:47 (libgrpc.so.7+0x12bfbc)
    #3 Thread /home/madeleineth/grpc/src/core/lib/gprpp/thd_posix.cc:142 (libgrpc.so.7+0x12bfbc)

  Thread T4 'grpcpp_sync_ser' (tid=10700, running) created by main thread at:
    #0 pthread_create ??:? (server+0x42e0c6)
    #1 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*) /home/madeleineth/grpc/src/core/lib/gprpp/thd_posix.cc:73 (libgrpc.so.7+0x12bfbc)
    #2 grpc_core::(anonymous namespace)::ThreadInternalsPosix* grpc_core::New<grpc_core::(anonymous namespace)::ThreadInternalsPosix, char const*&, void (*&)(void*), void*&, bool*>(char const*&, void (*&)(void*), void*&, bool*&&) /home/madeleineth/grpc/./src/core/lib/gprpp/memory.h:47 (libgrpc.so.7+0x12bfbc)
    #3 Thread /home/madeleineth/grpc/src/core/lib/gprpp/thd_posix.cc:142 (libgrpc.so.7+0x12bfbc)
    #4 __libc_start_main /build/glibc-B9XfQf/glibc-2.28/csu/../csu/libc-start.c:308 (libc.so.6+0x2409a)

SUMMARY: ThreadSanitizer: data race ??:? in free
```
 
### Anything else we should know about your project / environment?

The machine I tested this on does not have any other gRPC or protobuf versions installed, and there is nothing suspicious in `ldd client` or `ldd server`.

I have seen a similar issue with v1.17.1 on Ubuntu 18.04 with Clang 6 and static libraries.

**Comment by madeleineth** [comment]
Poking this to keep it from getting closed. I'm trying to replicate it against HEAD, but I ran into some entirely different threading issues. I will update the thread once I have more information.

**Comment by yashykt** [comment]
Thanks @madeleineth ! Please run it from head. It would also be good to have better traces.

**Comment by yashykt** [comment]
This was probably fixed by #19531
